src/Veldrid/OpenGL/OpenGLGraphicsDevice.cs

```diff
@@ -766,15 +770,37 @@ protected override void UpdateBufferCore(DeviceBuffer buffer, uint bufferOffsetI
             uint mipLevel,
             uint arrayLayer)
         {
-            StagingBlock sb = _stagingMemoryPool.Stage(source, sizeInBytes);
-            _executionThread.Run(() =>
-            {
-                fixed (byte* dataPtr = &sb.Array[0])
-                {
-                    _commandExecutor.UpdateTexture(texture, (IntPtr)dataPtr, x, y, z, width, height, depth, mipLevel, arrayLayer);
-                    sb.Free();
-                }
-            });
+            StagingBlock textureData = _stagingMemoryPool.Stage(source, sizeInBytes);
+            StagingBlock argBlock = _stagingMemoryPool.GetStagingBlock(UpdateTextureArgsSize);
+            ref UpdateTextureArgs args = ref Unsafe.AsRef<UpdateTextureArgs>(argBlock.Data);
+            args.Texture = texture;
+            args.Data = (IntPtr)textureData.Data;
+            args.X = x;
+            args.Y = y;
+            args.Z = z;
+            args.Width = width;
+            args.Height = height;
+            args.Depth = depth;
+            args.MipLevel = mipLevel;
+            args.ArrayLayer = arrayLayer;
+
+            _executionThread.UpdateTexture(argBlock.Id, textureData.Id);
+        }
+
+        public static readonly uint UpdateTextureArgsSize = (uint)Unsafe.SizeOf<UpdateTextureArgs>();
+
+        public struct UpdateTextureArgs
```

Comment:
Same, it can probably be private

Reply:
I think I left it public because I was experimenting with this idea
>We could go even further and change OpenGLCommandExecutor.UpdateTexture to accept a reference to UpdateTextureArgs instead of separate parameters to avoid extra copying